MNT: Reorganize non-constant global statics into structs

ngoldbaum opened the pull request:
This reorganizes most of the mutable static globals in numpy into four structs exposed to the internal API via `multiarraymodule.h`

* `npy_interned_str` for interned strings
* `npy_static_pydata` for immutable PyObjects that are initialized during module initialization
* `npy_static_cdata` for immutable C data that is initialized during module initialization
* `npy_ma_thread_unsafe_state` for state stored in a thread-unsafe manner

With the goal of refactoring the items in `npy_ma_thread_unsafe_state` to be thread-safe in followup PRs. See also the tracking issue for items that still need to be fixed.

ngoldbaum committed:
MNT: move cached imports into a global struct

ngoldbaum committed:
MNT: move interned strings into a single global struct

ngoldbaum committed:
MNT: move cpu dispatch registry into global data struct

ngoldbaum committed:
MNT: move ndarray.__array_*__ references to global data struct

ngoldbaum committed:
MNT: move sys.flags.optimize cache to global data struct

ngoldbaum committed:
MNT: set up tuple for truediv in global data struct

ngoldbaum committed:
MNT: move unpack_bits LUT into global static struct

ngoldbaum committed:
MNT: move references to int(1) and int(0) to global static struct

ngoldbaum committed:
MNT: move initialization of global ArrayMethods to module initialization

ngoldbaum committed:
MNT: move initialization of global tuples to global data struct

ngoldbaum committed:
MNT: move default extobj contextvar to global data dict

ngoldbaum committed:
MNT: move PyArray_SetStringFunction internals into global data struct

ngoldbaum committed:
BUG: remove questionable static initialization of an array object

ngoldbaum committed:
MNT: split global data struct into two structs

ngoldbaum commented:
I ran the full benchmark suite. It looks like there are some performance changes but it's kind of obnoxious to pick out reproducible changes from the big list `spin bench -c` spits out.

So far I've found:

```
| Change   | Before [a2d19725] <main>   | After [95a592b5] <global-state-refactor>   |   Ratio | Benchmark (Parameter)                                           |
|----------|----------------------------|--------------------------------------------|---------|-----------------------------------------------------------------|
| +        | 3.22±0.01μs                | 5.82±0.03μs                                |    1.81 | bench_ufunc.NDArrayAsType.time_astype(('float32', 'complex64')) |
| +        | 3.44±0.03μs                | 5.87±0.02μs                                |    1.71 | bench_ufunc.NDArrayAsType.time_astype(('float64', 'complex64')) |
```

see the full list here: https://gist.github.com/ngoldbaum/7e8ee9a129a96a32536f228e4214018b

I suspect the majority is noise. I'll try to pare it down to just the real performance changes and see if I can understand what's going on.

ngoldbaum commented:
I tried comparing `main` with itself on the full benchmark suite using `spin bench -c` and saw no changes, so the performance changes seen above certainly aren't noise like I said earlier. Will try to dig in to understand why some of the differences end up being random large performance regressions or improvements depending on the test run...

mattip commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -4795,142 +4790,196 @@ set_flaginfo(PyObject *d)
-    if (npy_ma_str_current_allocator == NULL) {
+    // this is module-level global heap allocation, it is currently
+    // never freed
+    npy_ma_str = PyMem_Calloc(sizeof(npy_ma_str_struct), 1);
```
Should this (and the others like it) check that it is not called twice?

mattip commented:
Looks like a move in the right direction to me. I wonder if the performance changes if you statically allocate the structs rather than calling `PyMem_Calloc`. Does the dynamic allocation affect the size of the shared object or is it too small to notice?

ngoldbaum committed:
MNT: make data structs static instead of heap-allocated

ngoldbaum commented:
> I wonder if the performance changes if you statically allocate the structs rather than calling PyMem_Calloc.

Good point! It does seem to help, at least on Linux. There are still some heap allocations though, I still need to look closer at whether they can be made static.

ngoldbaum committed:
MNT: add PyArrayMethodObject caches to static data struct

ngoldbaum committed:
MNT: move some thread-unsafe state in thread-unsafe state struct

ngoldbaum commented:
So the latest version of this PR drops moving things into the `npy_ma_state` struct to avoid heap allocations entirely. I think that does improve performance but it's difficult for me to tell which benchmarks give useful results and which don't.

Earlier I said I compared `main` with itself using `spin bench`, but I did that using `spin bench -c main` with `main` checked out, and it looks like there is some state that asv persists between runs on the same commit hash, because when I do the same thing but with two identical commits with different hashes (generated by doing `git commit --ament` with no changes to the working directory), I see random changes in the results for some benchmarks:

`` `
$ spin bench -c f3643a2d85ef4c18769608e111d465ce4a778474 -v -t "bench_linalg.Linalg"
... eliding a bunch of output ...
| Change   | Before [f3643a2d]    | After [9e40ee24] <global-state-refactor>   |   Ratio | Benchmark (Parameter)                      |
|----------|----------------------|--------------------------------------------|---------|--------------------------------------------|
| +        | 783±3μs              | 1.45±0.7ms                                 |    1.85 | bench_linalg.Linalg.time_svd('int16')      |
| +        | 32.6±0.05μs          | 45.9±10μs                                  |    1.41 | bench_linalg.Linalg.time_det('complex64')  |
| -        | 35.5±0.04μs          | 32.5±0.03μs                                |    0.92 | bench_linalg.Linalg.time_det('float32')    |
| -        | 2.18±0.06ms          | 785±0.9μs                                  |    0.36 | bench_linalg.Linalg.time_svd('float64')    |
| -        | 2.20±0.04ms          | 792±1μs                                    |    0.36 | bench_linalg.Linalg.time_svd('int64')      |
| -        | 109±70μs             | 35.9±0.2μs                                 |    0.33 | bench_linalg.Linalg.time_det('complex128') |
```

Here f3643a2d is the most recent commit in this PR and 9e40ee24 is the commit I generated with a no-op `git commit --amend`.

I'm going to start a full benchmark run like this to determine which benchmarks have random results and I will ignore those for determining whether or not this PR has a performance impact.

rgommers commented:
I see what the bug is there. `param` definition uses `set`, and the run-to-run variation makes `asv` misbehave. This diff fixes the problem:
```diff
diff --git a/benchmarks/benchmarks/bench_linalg.py b/benchmarks/benchmarks/bench_linalg.py
index 3077357237..f3eb819c18 100644
--- a/benchmarks/benchmarks/bench_linalg.py
+++ b/benchmarks/benchmarks/bench_linalg.py
@@ -72,7 +72,7 @@ def time_tensordot_a_b_axes_1_0_0_1(self):
 
 
 class Linalg(Benchmark):
-    params = set(TYPES1) - set(['float16'])
+    params = sorted(list(set(TYPES1) - set(['float16'])))
     param_names = ['dtype']
 
     def setup(self, typename):
```

ngoldbaum commented:
Nice! That makes sense. This is just one benchmark I could run quickly to prove there's an issue, I'll take a look at the full list of randomly changing tests to see if there are similar problems.

seberg reviewed (commented):
Left some comments.  I am still a bit saddened by having these giant init functions and wonder if we shouldn't do a local init pattern instead for some things.

I may look into some maintanence here in general, I think we can invent some cuter patterns (even if it adds a bit of complexity in the helpers).
I also think we should just drop the `_ma_` for "multiarray", it is a leftover of when multiarray and ufunc were two modules.
I.e. we could shorten it to `npy_static.<...>` even.

ngoldbaum commented on numpy/_core/src/multiarray/multiarraymodule.h:
```diff
@@ -1,26 +1,228 @@
+     *
+     * so the initialization is thread-safe and the only possibile lock
+     * contention happens before the cache is initialized, not on every single
+     * read.
```
Yeah, see https://github.com/numpy/numpy/pull/26430 for that. After this gets merged my next step is to integrate that into the argparse cache and this cache to make them thread-safe.

seberg replied:
So, I think the question remains a bit how many of these we actually want to centralize.  For strings and imports, I think it makes sense for sure.  But for the `ArrayMethods` used in the cast for example I would prefer to keep the code where it is.
And at that point, I am not sure it would be better to call the function from the main init function or rather just lock there.

ngoldbaum replied:
> And we will also do this for the argparse things, or do you aim to never do any init locally to avoid having to remember to explicitly lock there?

But yes, the argparse cache macro will also be updated like in #26430, the only major difference is the global lock I added there to handle locking during cache initialization would live in the global state struct. But maybe not if you're sour on that idea and it can just live in a spot local to the module like in #26430 as-is.

seberg replied:
I just meant that unless we have no localized init, there is maybe little reason to avoid it.

ngoldbaum commented on numpy/_core/src/multiarray/multiarraymodule.h:
```diff
@@ -1,26 +1,228 @@
+    union {
+        npy_uint8  bytes[8];
+        npy_uint64 uint64;
+    } unpack_lookup_big[256];
```
> Also, this table for example is truly static even with subinterpreters. The only issue is initialization.

Yup, everything in this struct is static after module initialization. I guess if we ever supported subinterpreters someone could make this be initialized once for all subinterpreters or just do it every time. I don't think it makes a ton of difference...

ngoldbaum commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -4795,147 +4790,191 @@ set_flaginfo(PyObject *d)
-    npy_ma_str___dlpack__ = PyUnicode_InternFromString("__dlpack__");
-    if (npy_ma_str___dlpack__ == NULL) {
+    npy_ma_str.__dlpack__ = PyUnicode_InternFromString("__dlpack__");
+    if (npy_ma_str.__dlpack__ == NULL) {
```
> I also think we should move this into it's own file.

Sorry, just to clarify - what should be moved into its own file? Filling all the static structs? Or just the string interning?

For the former, fair enough, that occured to me, and it would separate multiarraymodule.h from the global statics.

seberg replied:
I would say all of the static ones in one file maybe?  Seemed worthwhile to try to use this opportunity to clean things up a bit nicer (but I am happy to help with it!)

ngoldbaum committed:
MNT: apply sebastian's refactoring suggestions

ngoldbaum committed:
MNT: move static data structs into their own file

ngoldbaum committed:
MNT: Add more global state I missed to the thread_unsafe_state struct

ngoldbaum commented:
I just split out the static data structs into their own file. I left the thread unsafe state struct in multiarraymodule.h since I don't think it makes sense to live in the new header and I like how it makes it clear when code accesses thread unsafe state and also to make it clearer where the state is defined. To justify that choice: I completely missed some global state on the first pass of this PR and added it to the thread unsafe state struct in this pass.

I looked at initializing e.g. the ArrayMethod objects used in the casts in a module init and I felt like it wasn't any clearer to initialize "locally" in an initialization function inside the same file as where the global is used or to initialize it along with all the other cached globals and it sort of feel like a nice pattern to centralize this stuff. All IMO of course.

I've been looking at benchmarking closely the past few days and I think all the benchmark results I shared earlier for this PR are noise. Many of the things that show up are due to bugs I fixed in #26637, #26638, and #26639. The rest I suspect are due to jitter on the laptop I was using to run the benchmarks as well as asv caching results. So far I haven't been able to find a single performance change reported by asv that is reproduce outside of the asv environment or that persists if I purge the asv results database or make a no-op empty commit to test with.

I'm starting an asv run with higher-than-default settings for `rounds`, `repeat`, and `sample_time` to spin over the weekend, which should hopefully leave me with a more manageable list of performance changes to sift through, if any.

mattip commented on numpy/_core/src/multiarray/npy_static_data.c:
```diff
@@ -0,0 +1,269 @@
+NPY_NO_EXPORT int
+intern_strings(void)
+{
+    npy_interned_str.current_allocator = PyUnicode_InternFromString("current_allocator");
```
I think there are ways to make this more compact, but maybe not worth the effort. A different paradigm (for instance, making `npy_interned_str` an array of pointers, and using enum-ed constants like `CURRENT_ALLOCATOR` to index the array) might allow checking that all the values/fields have been initialized. Here I guess you could do a loop over the `sizeof(npy_interned_str)/sizeof(char *)` at the end of the function and check for `NULL`

Likewise, it would be nice to check `npy_static_pydata` for NULL values after initialization.

ngoldbaum replied:
See the latest commit which adds a function that verifies all entries in the two structs are non-null once module initialization is complete.

It unfortunately can't happen right after initializing the structs in this file here because some items (e.g. the reference to ndarray's `__array_ufunc__` implementation) cannot be created until relatively late during module initialization, but other things are needed early.

mattip replied:
Thanks, I think that check will prevent future problems when changing these structs.

ngoldbaum committed:
MNT: verify all entries in npy_interned_str and npy_static_pydata are non-null

seberg reviewed (commented):
Had a look through with some nitpicky comments.  Thanks for reorganizing into a new file, and the macro's for strings and imports are also at least much more compact!

Overall, I think this is good to go for me, unfortunately it might create merge conflicts pretty quickly, but I guess we'll just have to deal with them (for backporting)?

(There are the nitpicky comments and a merge conflict, but I can also just make a pass and merge if you think it's ready and it is clear that we are merging now.)

ngoldbaum commented on numpy/_core/src/multiarray/multiarraymodule.h:
```diff
@@ -1,26 +1,114 @@
+     *      struct {
+     *          atomic_int initialized;
+     *          PyObject *value;
+     *      }
```
You need to use at least one atomic load otherwise it's possible that the compiler might reorder things like:

```
// original code
if (flag) {
    do_some_work();
    do_more_work();
}

// after optimization
if (flag) {
    do_some_work();
}
if (flag) {
    do_more_work()
}
```

because data races are UB. In this case it would just lead to a memory leak. For the argparse cache I think it could cause two threads to simultaneously run `initialize_keywords`, corrupting the cache.

seberg replied:
Yeah, I suspected that much, was just wondering if the `atomic_int` is strictly necessary (or there is a way to do an `atomic_load(value)`, I guess).

ngoldbaum replied:
Ah no it's not necessary, I can do it with an atomic load instead.

ngoldbaum commented on numpy/_core/src/multiarray/multiarraymodule.h:
```diff
@@ -1,26 +1,114 @@
+     * This should be removed, see gh-26576
+     */
+    PyObject *PyArray_StrFunction;
+    PyObject *PyArray_ReprFunction;
```
Yeah I'm happy managing the merge conflicts for this stuff I'm actively removing :)

ngoldbaum committed:
Apply suggestions from code review

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

ngoldbaum committed:
MAINT: apply more of Sebastian's suggestions

ngoldbaum commented on numpy/_core/src/umath/ufunc_object.c:
```diff
@@ -698,16 +698,13 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
+                    goto fail;
                 }
+                ((PyArrayObject_fields *)zero_arr)->flags |= (
+                        NPY_ARRAY_WAS_PYTHON_INT|NPY_ARRAY_WAS_INT_AND_REPLACED);
```
Done in the latest version

ngoldbaum commented on numpy/_core/src/multiarray/npy_static_data.c:
```diff
@@ -0,0 +1,262 @@
+     *
+     * Note that some entries are initialized elsewhere. Care
+     * must be taken to ensure all entries are initialized during
+     * module initialization and immutable thereafter.
```
See latest commit.

ngoldbaum commented:
It seems gcc generates code that seg faults if you try to write to a field of a const static struct via a non-const pointer so unfortunately I don't think I can easily make the structs const. I could have two sets of structs - const structs used only to read from that are filled in at the end of initialization and non-const structs filled in during initialization but that didn't seem worth the additional complexity to me.

ngoldbaum commented:
I think this is ready to merge now. Since this will definitely conflict with the `set_string_function` PR, why don't we merge that one first, I'll fix the conflicts here, and then we can merge this one?

Also if there are worries about generating lots of conflicts to the 2.0 maintenance branch I'd be happy to hold off on merging this until we're a little closer to the 2.1 maintenance branch being created. I can continue to update this as I make more things thread-safe.

ngoldbaum commented:
We discussed this at the community meeting and agreed to merge it now, so I'm pulling this in. Thanks all for the reviews!
